synedrion/src/uint/signed.rs

```diff
@@ -332,76 +435,447 @@ where
 }
 
-impl<T: UintLike> Add<&Signed<T>> for Signed<T> {
+impl<T> Add<&Signed<T>> for Signed<T>
+where
+    T: Integer + crypto_bigint::Bounded,
+{
     type Output = Self;
     fn add(self, rhs: &Self) -> Self::Output {
         self.checked_add(rhs).unwrap()
     }
 }
 
-impl<T: UintLike> Sub<Signed<T>> for Signed<T> {
+impl<T> CheckedSub<Signed<T>> for Signed<T>
+where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Integer,
+{
+    /// Performs subtraction that returns `None` instead of wrapping around on underflow.
+    /// The bound of the result is the bound of `self` (lhs).
+    fn checked_sub(&self, rhs: &Signed<T>) -> CtOption<Self> {
+        self.value.checked_sub(&rhs.value).and_then(|v| {
+            let signed = Signed::new_positive(v, self.bound);
+            if let Some(signed) = signed {
+                CtOption::new(signed, 1u8.into())
+            } else {
+                CtOption::new(Signed::default(), 0u8.into())
+            }
+        })
+    }
+}
+
+impl<T> Sub<Signed<T>> for Signed<T>
+where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Encoding + Integer,
+{
     type Output = Self;
     fn sub(self, rhs: Self) -> Self::Output {
-        self.checked_add(&-rhs).unwrap()
+        self.checked_add(&-rhs).expect("Invalid subtraction")
     }
 }
 
-impl<T: UintLike> Sub<&Signed<T>> for Signed<T> {
+impl<T> Sub<&Signed<T>> for Signed<T>
+where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Encoding + Integer,
+{
     type Output = Self;
     fn sub(self, rhs: &Self) -> Self::Output {
-        self.checked_add(&-rhs).unwrap()
+        self.checked_add(&-rhs).expect("Invalid subtraction")
     }
 }
 
-impl<T: UintLike> Mul<Signed<T>> for Signed<T> {
+impl<T> Mul<Signed<T>> for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded + ConditionallySelectable,
+{
     type Output = Self;
     fn mul(self, rhs: Self) -> Self::Output {
         self.checked_mul(&rhs).unwrap()
     }
 }
 
-impl<T: UintLike> Mul<&Signed<T>> for Signed<T> {
+impl<T> Mul<&Signed<T>> for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded + ConditionallySelectable,
+{
     type Output = Self;
     fn mul(self, rhs: &Self) -> Self::Output {
         self.checked_mul(rhs).unwrap()
     }
 }
 
-impl<T: UintLike> core::iter::Sum for Signed<T> {
+impl<T> core::iter::Sum for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded,
+{
     fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
         iter.reduce(|x, y| x.checked_add(&y).unwrap())
             .unwrap_or(Self::default())
     }
 }
 
-impl<'a, T: UintLike> core::iter::Sum<&'a Self> for Signed<T> {
+impl<'a, T> core::iter::Sum<&'a Self> for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded,
+{
     fn sum<I: Iterator<Item = &'a Self>>(iter: I) -> Self {
         iter.cloned().sum()
     }
 }
+
+// @reviewers: Implementing `PartialEq`/`Eq` this way disregards the bound when it comes to equality,
```

Comment:
The reason it had been behaving like that before was that `PartialEq` had been only used in tests, so the comparison of bounds served as an additional assertion. Now that I think about it, perhaps the `PartialEq` impl should be made test-only (regardless of whether we decide to compare bounds or not) to prevent potential misuse.

Reply:
`PartialEq` is definitely used, e.g. [here](https://github.com/dvdplm/synedrion/blob/b02782ad9c987d45ffba20ba839ad25b300085fc/synedrion/src/cggmp21/sigma/fac.rs#L161) so we have to make a call here. The conservative option seems to back out my addition here and revert back to deriving it. Works for you?

Reply:
Yes, I think it is better to keep the strict behavior unless we specifically need to relax it.

Reply:
Done!